pkg/query-service/auth/auth.go

```diff
@@ -207,65 +209,121 @@ type RegisterRequest struct {
 	Email       string `json:"email"`
 	Password    string `json:"password"`
 	InviteToken string `json:"token"`
+
+	// reference URL to track where the register request is coming from
+	SourceUrl string `json:"sourceUrl"`
 }
 
-// Register registers a new user. For the first register request, it doesn't need an invite token
-// and also the first registration is an enforced ADMIN registration. Every subsequent request will
-// need an invite token to go through.
-func Register(ctx context.Context, req *RegisterRequest) *model.ApiError {
+func RegisterFirstUser(ctx context.Context, req *RegisterRequest) (*model.User, *model.ApiError) {
 
-	zap.S().Debugf("Got a register request for email: %v\n", req.Email)
+	if req.Email == "" {
+		return nil, model.BadRequest(model.ErrEmailRequired{})
+	}
 
-	// TODO(Ahsan): We should optimize it, shouldn't make an extra DB call everytime to know if
-	// this is the first register request.
-	users, apiErr := dao.DB().GetUsers(ctx)
+	if req.Password == "" {
+		return nil, model.BadRequest(model.ErrPasswordRequired{})
+	}
```

Comment:
Can these types of validations be moved to the initial request parsing itself? I noticed them in multiple places.